sympy/physics/control/lti.py

```diff
@@ -1422,3 +1445,563 @@ def _eval_rewrite_as_TransferFunction(self, num, den, **kwargs):
+    [  -----     ----------]
+    [  s + 2      2        ]
+    [            s  + s + 1]
+    [                      ]
+    [    3          -12    ]
+    [  -----       -----   ]
+    [  s + 2       s + 2   ]
+    >>> pprint(tfm_2, use_unicode=False) # State of tfm_2 is unchanged after substitution
+    [   a + s           -3       ]
+    [ ----------       -----     ]
+    [  2               s + 2     ]
+    [ s  + s + 1                 ]
+    [                            ]
+    [ 4                          ]
+    [p  - 3*p + 2      -a - s    ]
+    [------------    ----------  ]
+    [   p + s         2          ]
+    [                s  + s + 1  ]
+    [                            ]
+    [                 4          ]
+    [     3        - p  + 3*p - 2]
+    [   -----      --------------]
+    [   s + 2          p + s     ]
+
+    ``subs()`` also supports multiple substitutions.
+
+    >>> tfm_2.subs({p: 2, a: 1})  # substituting p with 2 and a with 1
+    TransferFunctionMatrix(((TransferFunction(s + 1, s**2 + s + 1, s), TransferFunction(-3, s + 2, s)), (TransferFunction(12, s + 2, s), TransferFunction(-s - 1, s**2 + s + 1, s)), (TransferFunction(3, s + 2, s), TransferFunction(-12, s + 2, s))))
+    >>> pprint(_, use_unicode=False)
+    [  s + 1        -3     ]
+    [----------    -----   ]
+    [ 2            s + 2   ]
+    [s  + s + 1            ]
+    [                      ]
+    [    12        -s - 1  ]
+    [  -----     ----------]
+    [  s + 2      2        ]
+    [            s  + s + 1]
+    [                      ]
+    [    3          -12    ]
+    [  -----       -----   ]
+    [  s + 2       s + 2   ]
+
+    Users can reduce the ``Series`` and ``Parallel`` elements of the matrix to ``TransferFunction`` by using
+    ``doit()``.
+
+    >>> tfm_6 = TransferFunctionMatrix([[Series(tf_3, tf_4), Parallel(tf_3, tf_4)]])
+    >>> tfm_6
+    TransferFunctionMatrix(((Series(TransferFunction(3, s + 2, s), TransferFunction(-a + p, 9*s - 9, s)), Parallel(TransferFunction(3, s + 2, s), TransferFunction(-a + p, 9*s - 9, s))),))
+    >>> pprint(tfm_6, use_unicode=False)
+    [ -a + p   3     -a + p     3  ]
+    [-------*-----  ------- + -----]
+    [9*s - 9 s + 2  9*s - 9   s + 2]
+    >>> tfm_6.doit()
+    TransferFunctionMatrix(((TransferFunction(-3*a + 3*p, (s + 2)*(9*s - 9), s), TransferFunction(27*s + (-a + p)*(s + 2) - 27, (s + 2)*(9*s - 9), s)),))
+    >>> pprint(_, use_unicode=False)
+    [    -3*a + 3*p     27*s + (-a + p)*(s + 2) - 27]
+    [-----------------  ----------------------------]
+    [(s + 2)*(9*s - 9)       (s + 2)*(9*s - 9)      ]
+
+    See Also
+    ========
+
+    TransferFunction, Series, Parallel, Feedback
+
+    """
+    def __new__(cls, arg):
+
+        expr_mat_arg = []
+        var = arg[0][0].var
+        for row_index, row in enumerate(arg):
+            temp = []
+            for col_index, element in enumerate(row):
+                if not isinstance(element, (TransferFunction, Series, Parallel)):
```

Comment:
These should have a common superclass that you can check for so you don't need an explicit list of types in a situation like this.

Reply:
Can you give an example?
I am very clear about what you are trying to say, but what should I name the class?

Reply:
The lazy is to call it something like `TransferFunctionBase`. That's clear to programmers but makes no mathematical sense to users. For users it's better if you have a meaningful name for the kinds of objects that can go into a `TransferFunctionMatrix`.

It would be the superclass of the others and would define the common methods like `to_expr`. The name should match up with whatever these conceptually represent if possible.

Reply:
@namannimmo10, what do you suggest?

Reply:
We did consider introducing a class called `TransferFunctionExpr` last year so that we could subclass `TransferFunction`, `Series`, and `Parallel` from it and make `__add__`, `__mul__` overloaded for `TransferFunctionExpr` with `Series` and `Parallel`. It could somewhat mimic the `MatrixExpr` superclass in the matrices module:
https://github.com/sympy/sympy/blob/200788d6d52657ae36d21737ed780471ec38be27/sympy/matrices/expressions/matexpr.py#L35
Finally, though, didn't find any harm in overloading operators in Series and Parallel classes. 

I'm not sure if we really need to introduce such a class even now, but since @oscarbenjamin is well versed with the matrices module, maybe it _is_ needed to structure the code. Can this be done later on?

Reply:
I wouldn't try to emulate the design of the matrices module. The reason for wanting a superclass is not so the subclasses can share code but rather:

1. Be able to define in a meaningful way what kinds of objects can be used in what situations
2. Define the interface that is expected of those objects
3. Document that clearly for users and contributors